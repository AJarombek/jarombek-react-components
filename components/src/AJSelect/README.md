### Overview

Reusable & custom <select> element.  Has a label and a dropdown to pick options from. 

**Use cases:**

```javascript
import React from 'react';
import { AJSelect } from 'jarombek-react-components';

const CustomComponent = () => 
  <AJSelect 
    options={[{content: 'Bound 2', value: 'bound2'}]}
    defaultOption={0}
    disabled={false}
  />
```

### Files

| Filename                   | Description                                                             |
|----------------------------|-------------------------------------------------------------------------|
| `AJSelect.js`              | The reusable `AJSelect` component.                                      |
| `index.js`                 | Index file which exports the `AJSelect` component.                      |
| `styles.js`                | JSS styles for the `AJSelect` component.                                |
