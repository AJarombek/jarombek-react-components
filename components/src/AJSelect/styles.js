/**
 * Stylesheet for the AJSelect component using JSS.
 * Author: Andrew Jarombek
 * Date: 8/27/2020
 */

import globalStyles from '../globalStyles';

export default {
  ajSelect: {
    display: 'flex'
  },
  ajSelectArrow: {
    ...globalStyles.elegantIcons,
    paddingLeft: 10
  },
  ajSelectLabel: {
    display: 'flex',
    position: 'relative',
    cursor: 'pointer',
    height: 30,
    padding: 5,
    border: '1px solid #BBB',
    borderRadius: 3,
    alignItems: 'center'
  },
  ajSelectDisabled: {
    backgroundColor: '#EEE',
    cursor: 'not-allowed'
  },
  ajSelectList: {
    position: 'absolute',
    cursor: 'pointer',
    margin: 0,
    padding: 0,
    top: 50,

    '& > li': {
      listStyleType: 'none',
      padding: '10px 6px',
      textAlign: 'center',

      '&:hover': {
        backgroundColor: '#d2d2d2'
      }
    }
  }
};
