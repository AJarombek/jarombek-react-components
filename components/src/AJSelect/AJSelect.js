/**
 * Reusable select element which has the same functionality as the HTML <select> element.
 * @author Andrew Jarombek
 * @since 8/26/2020
 */

import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {createUseStyles} from 'react-jss';
import classnames from 'classnames';

import styles from './styles';

const useStyles = createUseStyles(styles);

const AJSelect = ({
  title,
  options,
  defaultOption=0,
  onClickSelect,
  onClickListOption,
  disabled=false,
  className
}) => {
  const classes = useStyles();
  const [selected, setSelected] = useState(defaultOption);
  const [isOpen, setIsOpen] = useState(false);

  const handleOnClickListOption = (option, index) => {
    onClickListOption(option);
  };

  return (
    <div className={classnames('aj-select', classes.ajSelect)}>
      <div onClick={onClickSelect}>
        <div>{title}</div>
        <div className={classnames(
          classes.ajSelectArrow, isOpen ? classes.ajSelectOpen : classes.ajSelectClosed
        )}>
          {isOpen ? '\u0042' : '\u0043'}
        </div>
      </div>
      {isOpen && (
        <ul className={classes.ajSelectList}>
          {options.map((option, index) => (
            <li key={option.value} onClick={() => handleOnClickListOption(option, index)}>
              {option.content}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

AJSelect.propTypes = {
  title: PropTypes.node.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    content: PropTypes.node,
    value: PropTypes.object
  })).isRequired,
  defaultOption: PropTypes.number,
  onClickSelect: PropTypes.func,
  onClickListOption: PropTypes.func,
  disabled: PropTypes.bool,
  className: PropTypes.string
};

export default AJSelect;
