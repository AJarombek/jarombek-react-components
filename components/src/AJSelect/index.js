import AJSelect from './AJSelect';

export default AJSelect;
