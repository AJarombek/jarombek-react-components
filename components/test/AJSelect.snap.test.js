import React from 'react';
import AJSelect from '../src/AJSelect/AJSelect';
import renderer from 'react-test-renderer';

/**
 * Snapshot test for the {@link ../../components/AJSelect} component.
 * @author Andrew Jarombek
 * @since 8/29/2020
 */

it('renders correctly', () => {
  const tree = renderer
    .create(
      <AJSelect
        options={[
          { content: 'Item 1', value: 1 },
          { content: 'Item 2', value: 2 },
        ]}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
