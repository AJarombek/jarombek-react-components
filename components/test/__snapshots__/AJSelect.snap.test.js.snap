// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="aj-select ajSelect-0-2-1 aj-select-closed"
>
  <div
    className="ajSelectLabel-0-2-3"
    onClick={[Function]}
  >
    <div />
    <div
      className="ajSelectArrow-0-2-2"
    >
      C
    </div>
  </div>
</div>
`;
